### PR TITLE
[Feat] Add update GitOps repository workflow

### PR DESCRIPTION
**Pull request by saenyakorn**
## Why did you create this PR

-

## What did you do

- Add `update-gitops.yaml` workflow for open PR to cugetreg-gitops repo
- Add `gitops-mode` to choose between `PR` or `commit` 

## Workflow explain 
- When we push something to to `dev`, the action will calculate affected packages and push kustomize changed to gitios repo.
- When we push something to `beta`, the action will do pre-release thing, and then open the Pre-release PR. After we merge the PR, the action will bump the package version, build the docker images and open kustomize changed PR to GitOps repo.
- Similar to `beta`, when we push something to `main`, the action will do release thing, and the rest is the same as `beta`'s flow. 

## Demo

Workflow https://github.com/thinc-org/cugetreg/actions/runs/4106806503

Success Commit - https://github.com/thinc-org/cugetreg-gitops/commit/ac296cee9f9ada87fc91369eda675950bbcbd62f

Success opened PR - https://github.com/thinc-org/cugetreg-gitops/pull/6 

[https://dev.cugetreg.com](https://dev.cugetreg.com)

## Checklist

- [ ] Deploy a demo
- [ ] Check browsers compatibility
- [ ] Wrote coverage tests

### CODE REVIEW

File: .github/workflows/deploy-docker.yaml [hunk 1]
```diff
@@ -109,48 +119,30 @@ jobs:
         with:
           context: .
           file: apps/${{ matrix.packages.name }}/Dockerfile
-          tags: |
-            ${{ inputs.container-registry }}/${{ inputs.image-prefix }}/${{ matrix.packages.name }}:${{ matrix.packages.imageTag }}
+          tags: ${{ inputs.container-registry }}/${{ inputs.image-prefix }}/${{ matrix.packages.name }}:${{ matrix.packages.imageTag }}
           push: ${{ inputs.push }}
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new
```
Comment (suphon-t) [inline .github/workflows/deploy-docker.yaml:125]: Why not just use gha cache?

File: .github/workflows/deploy-docker.yaml [hunk 2]
```diff
@@ -109,48 +119,30 @@ jobs:
...
-  #       with:
-  #         title: Update Package versions
-  #         commit-message: Update Package versions
+      - name: Move Docker cache
+        if: steps.check-dockerfile.outputs.DOCKERFILE_EXISTS == 'true'
+        run: |
+          rm -rf /tmp/.buildx-cache
+          mv /tmp/.buildx-cache-new /tmp/.buildx-cache
+
+  update-gitops:
```
Comment (suphon-t) [inline .github/workflows/deploy-docker.yaml:133]: Should we move this to the release workflows and keep this one for build only?

File: .github/workflows/update-gitops.yaml [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      - name: Update kustomize configuration
+        run: |
+          for row in $(echo "$PACKAGES" | jq -r '.[] | @base64'); do
+            _jq() {
+              echo ${row} | base64 --decode | jq -r ${1}
+            }
+            NAME=$(_jq '.name')
+            IMAGE_TAG=$(_jq '.imageTag')
+            PREFIX=${{ inputs.container-registry }}/${{ inputs.image-prefix }}
+            bash -c "cd $NAME/overlays/${{ inputs.environment }} && kustomize edit set image $NAME=$PREFIX/$NAME:$IMAGE_TAG"
```
Comment (bombnp) [inline .github/workflows/update-gitops.yaml:67]: why put these two commands in `bash -c`?

File: .github/workflows/update-gitops.yaml [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            echo "${NAME}:${IMAGE_TAG} is updated"
+          done
+        env:
+          PACKAGES: ${{ inputs.packages }}
```
Comment (bombnp) [inline .github/workflows/update-gitops.yaml:71]: do u need to pass this as env? why not just use `${{ inputs.packages }}` in the for loop itself?

File: .github/workflows/update-gitops.yaml [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+      - name: Show Git Status
+        run: git status
+
+      - name: Prepare Pull Request body
+        id: pr-body
```
Comment (bombnp) [inline .github/workflows/update-gitops.yaml:77]: should add `if: ${{ inputs.mode == 'pr' }}`

File: .github/workflows/update-gitops.yaml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        run: |
+          echo "DATE=$(date +'%d/%m/%Y')" >> $GITHUB_OUTPUT
+          UPDATED_PACKAGES=$(echo "$PACKAGES" | jq -r '.[] | "- \(.name): \(.imageTag)"')
+          echo "UPDATED_PACKAGES<<EOF" >> $GITHUB_ENV
+          echo "$UPDATED_PACKAGES" >> $GITHUB_ENV
+          echo "EOF" >> $GITHUB_ENV
```
Comment (bombnp) [inline .github/workflows/update-gitops.yaml:83]: why do we use this as ENV instead of OUTPUT like DATE na?

**Review by bombnp — CHANGES_REQUESTED** [review]

File: .github/workflows/update-gitops.yaml (hunk 3 above)
Comment (suphon-t) [reply .github/workflows/update-gitops.yaml:67]: To spawn new shells to cd for each item, and not change the cwd of the main shell

File: .github/workflows/update-gitops.yaml (hunk 4 above)
Comment (saenyakorn) [reply .github/workflows/update-gitops.yaml:71]: if we put `${{ inputs.package }}` directly. I will be something like.

```
for row in $(echo "[
 {
  "name": "brah",
  "imageTag": "brah"
 }
]" | jq -r '.[] | @base64'); do
```

and broke the script. 

To enforce the value within a single line, we have to use env.

File: .github/workflows/update-gitops.yaml (hunk 6 above)
Comment (saenyakorn) [reply .github/workflows/update-gitops.yaml:83]: Because of this issue. https://github.com/octokit/request-action/issues/6 https://github.com/github/docs/issues/21529

Workaround https://trstringer.com/github-actions-multiline-strings/

TLDR; GitHut output fuck with multiple line output

File: .github/workflows/release.yaml [hunk 7]
```diff
@@ -140,12 +155,11 @@ jobs:
             "imageTag": "${{ github.event.inputs.version }}"
           }
         ]
+      environment: ${{ needs.release.outputs.environment }}
```
[the remark this answers is not in the document]
Comment (saenyakorn) [reply .github/workflows/release.yaml:158]: True

File: .github/workflows/release-dev.yaml [hunk 8]
```diff
@@ -68,10 +69,24 @@ jobs:
...
+            echo "ENVIRONMENT=prod" >> $GITHUB_OUTPUT
+          elif [[ ${{ github.ref_name  }} == "beta" ]]; then
+            echo "Deploy environment is beta"
+            echo "ENVIRONMENT=beta" >> $GITHUB_OUTPUT
+          elif [[ ${{ github.ref_name  }} == "dev" ]]; then
+            echo "Deploy environment is dev"
+            echo "ENVIRONMENT=dev" >> $GITHUB_OUTPUT
+          fi
+
       - name: Print Debug
```
[the remark this answers is not in the document]
Comment (saenyakorn) [reply .github/workflows/release-dev.yaml:90]: Well

File: .github/workflows/release-dev.yaml [hunk 9]
```diff
@@ -22,7 +22,7 @@ permissions:
 
 jobs:
   release:
-    name: Release dev
+    name: Release without versioning
```
[the remark this answers is not in the document]
Comment (saenyakorn) [reply .github/workflows/release-dev.yaml:25]: to what?

File: .github/workflows/deploy-docker.yaml (hunk 2 above)
Comment (saenyakorn) [reply .github/workflows/deploy-docker.yaml:133]: If that so, the `release.yaml` will have double `update-gitops.yaml` since GitHib action has no `OR` needs. Also, we may update GitOps if and only if we build docker. I think it make sense to put this here.

File: .github/workflows/deploy-docker.yaml (hunk 1 above)
Comment (saenyakorn) [reply .github/workflows/deploy-docker.yaml:125]: Link?

File: .github/workflows/deploy-docker.yaml [hunk 10]
```diff
@@ -55,6 +55,7 @@ jobs:
 
     outputs:
       DOCKERFILE_EXISTS: ${{ steps.check-dockerfile.outputs.DOCKERFILE_EXISTS }}
+      ENVIRONMENT: ${{ steps.get-deploy-environment.outputs.ENVIRONMENT }}
```
[the remark this answers is not in the document]
Comment (saenyakorn) [reply .github/workflows/deploy-docker.yaml:58]: well

File: .github/workflows/deploy-docker.yaml [hunk 11]
```diff
@@ -7,41 +7,41 @@ on:
...
         default: ${{ github.repository }}
-      target-ref:
+      target-gitops-ref:
         description: Target ref for updating deployment declaration
         type: string
-        default: main
-      target-directory:
-        description: Target directory for updating deployment declaration
+        default: ref/heads/master
+      gitops-mode:
```
[the remark this answers is not in the document]
Comment (saenyakorn) [reply .github/workflows/deploy-docker.yaml:29]: ok

File: .github/workflows/release-dev.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (suphon-t) [reply .github/workflows/release-dev.yaml:25]: To `release-without-versioning`?

File: .github/workflows/deploy-docker.yaml (hunk 1 above)
Comment (suphon-t) [reply .github/workflows/deploy-docker.yaml:125]: Just use `type=gha`, and GitHub Actions will handle everything automatically.

https://docs.docker.com/build/cache/backends/gha/

Comment (suphon-t) [reply .github/workflows/deploy-docker.yaml:125]: Also don't forget to add `mode=max` to `cache-to` to cache all layers.

File: .github/workflows/update-gitops.yaml (hunk 3 above)
Comment (bombnp) [reply .github/workflows/update-gitops.yaml:67]: ok

File: .github/workflows/release.yaml [hunk 12]
```diff
@@ -103,6 +105,20 @@ jobs:
...
+          if [[ ${{ github.ref_name  }} == "main" ]]; then
+            echo "Deploy environment is production"
+            echo "ENVIRONMENT=prod" >> $GITHUB_OUTPUT
+          elif [[ ${{ github.ref_name  }} == "beta" ]]; then
+            echo "Deploy environment is beta"
+            echo "ENVIRONMENT=beta" >> $GITHUB_OUTPUT
+          elif [[ ${{ github.ref_name  }} == "dev" ]]; then
+            echo "Deploy environment is dev"
+            echo "ENVIRONMENT=dev" >> $GITHUB_OUTPUT
+          fi
```
[the remark this answers is not in the document]
Comment (saenyakorn) [reply .github/workflows/release.yaml:120]: `update-gitops` can be triggered by `workflow_dispatch` (for release-non-prod). Which mean it does not know what environment will be used by using `github.ref_name`. We have to forward `environment` for the root of action.

**Review by bombnp — APPROVED** [review]

File: .github/workflows/release.yaml [hunk 13]
```diff
@@ -103,6 +113,20 @@ jobs:
           PACKAGES=$(echo ${{ steps.changesets.outputs.publishedPackages }} | jq -c 'map_values({name:.name,ref:"refs/tags/"+.version,imageTag:.version})')
```
Comment (suphon-t) [inline .github/workflows/release.yaml:113]:
```suggestion
          PACKAGES=$(echo ${{ steps.changesets.outputs.publishedPackages }} | jq -c 'map_values({name:.name,ref:"refs/tags/\(.version)",imageTag:.version})')
```

**Review by suphon-t — COMMENTED** [review]
One more thing. The old syntax seems to be broken.

**Review by suphon-t — APPROVED** [review]
